src/network/networkd-wifi-wpa_supplicant.c

```diff
@@ -0,0 +1,348 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "sd-bus.h"
+
+#include "bus-util.h"
+#include "networkd-link.h"
+#include "networkd-manager.h"
+#include "networkd-wifi-wpa_supplicant.h"
+#include "string-util.h"
+
+static int wpa_supplicant_get_network_properties_handler(sd_bus_message *m, void *userdata, sd_bus_error *ret_error) {
+        const char *name, *ssid = NULL;
+        const sd_bus_error *e;
+        Link *link = userdata;
+        int r;
+
+        assert(m);
+        assert(link);
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 0;
+
+        e = sd_bus_message_get_error(m);
+        if (e) {
+                log_link_full(link,
+                              sd_bus_error_has_name(e, "org.freedesktop.DBus.Error.UnknownMethod") ? LOG_DEBUG : LOG_ERR,
+                              sd_bus_error_get_errno(e),
+                              "Failed to get current wifi properties: %s",
+                              e->message);
+                return 0;
+        }
+
+        r = sd_bus_message_enter_container(m, 'v', "a{sv}");
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        r = sd_bus_message_enter_container(m, 'a', "{sv}");
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        for (;;)
+                if (ssid) {
+                        r = sd_bus_message_skip(m, "{sv}");
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                        if (r == 0)
+                                break;
+                } else {
+                        r = sd_bus_message_enter_container(m, 'e', "sv");
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                        if (r == 0)
+                                break;
+
+                        r = sd_bus_message_read(m, "s", &name);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+
+                        if (streq(name, "ssid")) {
+                                r = sd_bus_message_read(m, "v", "s", &ssid);
+                                if (r < 0)
+                                        return bus_log_parse_error(r);
+                        } else
+                                r = sd_bus_message_skip(m, "v");
+
+                        r = sd_bus_message_exit_container(m);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                }
+
+        r = sd_bus_message_exit_container(m);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        r = sd_bus_message_exit_container(m);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        /* It seems that ssid provided by wpa_supplicant is quoted... */
+        if (*ssid == '"')
+                ssid = strndupa(ssid + 1, strlen(ssid) - 2);
```

Comment:
not that it matters, but I'd probably move this to a helper function "wpa_supplicant_ssid_unquote()" or so, since the quoting is so weird and unnecesary..

(btw, what happens if the ssid contains a quote itself?)